src/main/java/graphql/execution/incremental/IncrementalCallState.java

```diff
@@ -0,0 +1,99 @@
+ * the main result is sent via a Publisher stream.
+ */
+@Internal
+public class IncrementalCallState {
+    private final AtomicBoolean incrementalCallsDetected = new AtomicBoolean(false);
+    private final Deque<IncrementalCall<? extends IncrementalPayload>> incrementalCalls = new ConcurrentLinkedDeque<>();
+    private final SingleSubscriberPublisher<DelayedIncrementalPartialResult> publisher = new SingleSubscriberPublisher<>();
+    private final AtomicInteger pendingCalls = new AtomicInteger();
+    private final LockKit.ReentrantLock publisherLock = new LockKit.ReentrantLock();
+
+    @SuppressWarnings("FutureReturnValueIgnored")
+    private void drainIncrementalCalls() {
+        IncrementalCall<? extends IncrementalPayload> incrementalCall = incrementalCalls.poll();
+
+        while (incrementalCall != null) {
+            incrementalCall.invoke()
+                    .whenComplete((payload, exception) -> {
+                        if (exception != null) {
+                            publisher.offerError(exception);
+                            return;
+                        }
+
+                        // The assigment of `remainingCalls` and `publisher.offer` need to be synchronized to ensure
+                        // `hasNext` is `false` precisely on the last event offered to the publisher.
+                        publisherLock.lock();
+                        final int remainingCalls;
+
+                        try {
+                            remainingCalls = pendingCalls.decrementAndGet();
+
+                            DelayedIncrementalPartialResult executionResult = newIncrementalExecutionResult()
+                                    .incrementalItems(Collections.singletonList(payload))
+                                    .hasNext(remainingCalls != 0)
+                                    .build();
+
+                            publisher.offer(executionResult);
+                        } finally {
+                            publisherLock.unlock();
+                        }
+
+                        if (remainingCalls == 0) {
+                            publisher.noMoreData();
+                        } else {
+                            // Nested calls were added, let's try to drain the queue again.
+                            drainIncrementalCalls();
+                        }
+                    });
+            incrementalCall = incrementalCalls.poll();
+        }
+    }
+
+    public void enqueue(IncrementalCall<? extends IncrementalPayload> incrementalCall) {
+        publisherLock.runLocked(() -> {
+            incrementalCallsDetected.set(true);
+            incrementalCalls.offer(incrementalCall);
+            pendingCalls.incrementAndGet();
+        });
+    }
+
+    public void enqueue(Collection<IncrementalCall<? extends IncrementalPayload>> calls) {
+        if (!calls.isEmpty()) {
+            calls.forEach(this::enqueue);
+        }
```

Comment:
why   `!calls.isEmpty())`

is this just and allAll kinda thing ??

Reply:
You're right, there's no need for this empty check. Let me remove it.